[PEP - Wheel Variant] Staging before push to upstream

DEKHTIARJonathan opened the pull request:
Rendered version: https://wheelnext.github.io/peps/pep-0815/

mgorny committed:
PEP 815: Raw conversion from the "Wheel Variants" Markdown

Done via:

```
m2r2 --no-underscore-emphasis pepxxx_wheel_variant_support.md
```

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Add proper PEP headers

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
PEP 815: fix hyperlink conflicts

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove incorrect lexers

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Copy assets over

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix pre-commit issues

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Sync PEP 440 ref removal

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Update Konstantin's name

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix pre-commit issues

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Update heading style per PEP 12

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix spacing around headings

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Restart with pandoc

The pandoc's conversion is much better.

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix headings again

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix appendix references

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix code blocks

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Reindent code blocks & fix syntax hl

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Diversify list enumerators

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Use definition list for glossary

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Use enumerated lists when we say "three"

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Reflow

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Provide alt text for images

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Make use of cross-references

Signed-off-by: Michał Górny <mgorny@quansight.com>

DEKHTIARJonathan committed:
WHLNXT render

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+        Konstantin Schütze <konstin@mailbox.org>,
+        Ralf Gommers <ralf.gommers@gmail.com>,
+        Andrey Talman <atalman@meta.com>,
+        Charlie Marsh <crmarsh416@gmail.com>,
```
Nit: Do you mind using `charlie.r.marsh@gmail.com` (if not `charlie@astral.sh`)?

mgorny replied:
I'm fine with either, I think I've copied that from your commit messages. Just tell me which one you prefer (or "suggest" a change).

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+
+The Python wheel packaging format uses
+:doc:`packaging:specifications/platform-compatibility-tags` to specify
+wheel’s supported environments based on Python version, ABI, and
```
```suggestion
a given wheel’s supported environments based on Python version, ABI, and
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+:doc:`packaging:specifications/platform-compatibility-tags` to specify
+wheel’s supported environments based on Python version, ABI, and
+platform (operating system, architecture, core system libraries).  These
+tags are not able to express features of modern hardware. This is
```
```suggestion
tags are unable to express features of modern hardware. This is
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+platform (operating system, architecture, core system libraries).  These
+tags are not able to express features of modern hardware. This is
+particularly challenging for the scientific computing, artificial
+intelligence (AI), machine learning (ML), and high-performance computing
```
```suggestion
intelligence (AI), machine learning (ML), and high-performance computing (HPC)
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+- An evolution of the wheel format called **Wheel Variant** that allows
+  wheels to be distinguished by hardware or software attributes.
+
+- A **variant provider plugin** interface allowing to dynamically detect
```
```suggestion
- A **variant provider plugin** interface that allows installers to dynamically detect
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+  platform attributes and select the most suitable wheel.
+
+The goal is to simplify the user experience to a familiar
+``pip install <package>``, while ensuring optimal performance and
```
```suggestion
``[uv] pip install <package>``, while ensuring optimal performance and
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+
+The Python packaging ecosystem has evolved to support increasingly
+diverse computing environments. The current software ecosystem often
+relies on platform specific features to pick which binaries are
```
```suggestion
relies on platform-specific features to determine which binaries are
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+The Python packaging ecosystem has evolved to support increasingly
+diverse computing environments. The current software ecosystem often
+relies on platform specific features to pick which binaries are
+compatible with a particular computer. Unfortunately the current wheel
```
```suggestion
compatible with a particular machine. Unfortunately the current wheel
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+diverse computing environments. The current software ecosystem often
+relies on platform specific features to pick which binaries are
+compatible with a particular computer. Unfortunately the current wheel
+format cannot adequately express the features of modern hardware. This
```
```suggestion
format cannot adequately express the diversity of features of present in modern hardware. This
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+limitation forces package authors into suboptimal distribution
+strategies and creates friction for users attempting to install
+performance-critical packages.
+
```
I think I'd add a paragraph here with a concrete example, i.e., bringing up the fact that packages have to be built for specific CUDA or ROCm versions which cannot be included in the wheel tag. As such, you find yourself needing to distribute many builds with identical versions and wheel tags, targeting very different hardware configurations.

mgorny replied:
Yeah, maybe, but we probably need to rewrite the next paragraph then as well.

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+details). Some methods include maintaining separate package indexes for
+different hardware configurations, bundling all potential variants into
+a single “mega-wheel” / “monolithic wheel” or using separate package
+names (``mypackage-gpu``, ``mypackage-cpu``). Each of these approaches
```
```suggestion
names (``mypackage-gpu``, ``mypackage-cpu``, etc.). Each of these approaches
```

charliermarsh commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2281 @@
+a single “mega-wheel” / “monolithic wheel” or using separate package
+names (``mypackage-gpu``, ``mypackage-cpu``). Each of these approaches
+has significant drawbacks, such as excessive binary size, dependency
+confusion, and inefficient dependency resolution, complex documentation,
```
```suggestion
confusion, inefficient dependency resolution, complex documentation,
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
  they indirectly do today.
```

charliermarsh replied:
(Can you expand on the last point? What do you mean by "indirectly"?)

mgorny replied:
@DEKHTIARJonathan, any chance you wrote that? Or have a clue what this is about?

rgommers replied:
See https://github.com/wheelnext/pep_xxx_wheel_variants/issues/54#issuecomment-3014031815. I don't think "they indirectly do today" is helpful indeed. More like "SciPy already ships wheels linked against different BLAS libraries on macOS, however they are not selectable by the user but keyed on different OS versions."

charliermarsh commented on peps/pep-0815.rst:
These are good examples.

charliermarsh commented on peps/pep-0815.rst:
```suggestion
chain attacks - more details on `PyTorch
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
a non-trivial problem to address which has forced the PyTorch team to create
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
a complete mirror of all their dependencies, and is one of the core
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
upgraded packages being reverted to the default variant on subsequent upgrades.
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
“Package names” as variants
```

charliermarsh replied:
Is it worth mentioning how this case does or doesn't accommodate cases like: I want to publish a package `foo` that supports using either `xgboost` or `xgboost-cpu`? Would I have to publish `foo` and `foo-cpu`, with a combination for every combination of dependencies?

mgorny replied:
https://github.com/wheelnext/peps/pull/7

charliermarsh commented on peps/pep-0815.rst:
```suggestion
``pip install jax`` (with no extra) leads to a broken installation for
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
`FlashAttention<https://github.com/Dao-AILab/flash-attention>`__ does
```

charliermarsh commented on peps/pep-0815.rst:
Isn't it also a security risk in that you lose some provenance because the `setup.py` is just downloading a wheel from the Internet?

mgorny replied:
I think my logic was "arbitrary code disqualifies it, so no need to bother more", but I'll add one more sentence.

mgorny replied:
https://github.com/wheelnext/peps/pull/7

charliermarsh commented on peps/pep-0815.rst:
There are lots of other examples of this, if helpful. vLLM does it, PyG does it (`pip install pyg-lib -f https://data.pyg.org/whl/torch-${TORCH}+${CUDA}.html`), etc.

charliermarsh commented on peps/pep-0815.rst:
Is this part true? I don't know that it feels justified by the above.

mgorny replied:
I guess you should know best — after all, doesn't `uv` special-case PyTorch already?

rgommers replied:
I think the quote I added about Spyder speaks to this. That said, I think that this is probably the least important compared to the burden on end users and package authors.

charliermarsh commented on peps/pep-0815.rst:
```suggestion
don’t take the time to read it, their install will likely be dysfunctional.
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
An additional limitation of this approach is that publishing a new release
```

charliermarsh commented on peps/pep-0815.rst:
Yeah, I think this should be a strong moment in the PEP. We should have quotes from developers of the above projects (e.g., vLLM).

rgommers replied:
I have 3 in a branch, will open a PR before the meeting. It was just hard to fit in without rewriting half the Motivation section - need to check if it's easier now.

rgommers replied:
Done in https://github.com/wheelnext/wheelnext/pull/110 separate, then can be synced to this PR.

charliermarsh commented on peps/pep-0815.rst:
```suggestion
- Maintains full backward compatibility with existing tools (i.e., non-variant aware installers, tools, and
  indexes remain unaffected).
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
- Enables automatic hardware-appropriate package selection.
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
- Supports the full spectrum of modern computing hardware.
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
- Provides a seamless and predictable experience for users, with little-to-no user input required.
```

charliermarsh commented on peps/pep-0815.rst:
This happened to me recently with vLLM, if you want a concrete example. They support an AVX-512 build, but I had to forgo AVX-512 support to make a minimally-compatible wheel.

charliermarsh commented on peps/pep-0815.rst:
```suggestion
    A string (up to 16 characters) added to the wheel filename to uniquely identify variants.
```

charliermarsh commented on peps/pep-0815.rst:
Because that "one dependency" is, presumedly, the Python package containing the provider itself?

mgorny replied:
Yes. It's basically "empty `requires` make no sense because then you don't have anything to call".

charliermarsh commented on peps/pep-0815.rst:
```suggestion
  value. Defaults to ``false``. If it is ``true``, the provider is
```

charliermarsh commented on peps/pep-0815.rst:
```suggestion
  considered optional and should not be used unless the user opts-in to
```

mgorny replied:
I don't think this is correct. The verb is "opt in", "opt-in" is the noun.

charliermarsh commented on peps/pep-0815.rst:
```suggestion
  incompatible. If it is ``false`` or missing, the provider is considered
```

mgorny committed:
Apply suggestions from code review

Co-authored-by: Charlie Marsh <crmarsh416@gmail.com>

mgorny committed:
Apply suggestion from @charliermarsh

Co-authored-by: Charlie Marsh <crmarsh416@gmail.com>

mgorny committed:
Include problems with depending on "package name" variants

https://github.com/wheelnext/peps/pull/6#discussion_r2579581623

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Extend risks for fetching via sdist

https://github.com/wheelnext/peps/pull/6#discussion_r2579584325

Signed-off-by: Michał Górny <mgorny@quansight.com>

jithunnair-amd commented on peps/pep-0815.rst:
```suggestion
    Wheel variants are a clear step in the right direction in this regard.
```

rgommers commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+PEP: 815
+Title: Wheel Variants - extending platform awareness
```
As just discussed, "awareness" might be a little off and it's about more than platform only.

Maybe a more accurate one would be something like: _Wheel Variants - hardware and build config specific wheels_?

mgorny committed:
Apply suggestion from @jithunnair-amd

Co-authored-by: Jithun Nair <37884920+jithunnair-amd@users.noreply.github.com>

emmatyping-nv committed:
Reduce size of PEP 815 images using oxipng

This commit reduces the size of the images associated with PEP 815 by
~40% each. The commandline used to reduce the size was `oxipng -o 6 --strip
safe --alpha $file`.

DEKHTIARJonathan committed:
Merge pull request #9 from wheelnext/pep-wheel-variants-reduce-image-sizes

Reduce size of variants PEP images using oxipng

emmatyping-nv commented on peps/pep-0815.rst:
"the packaging ecosystem version" is a bit unclear here, I'm not sure what is meant by it.

mgorny replied:
@DEKHTIARJonathan, that's probably you?

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
wheel characteristics beyond what existing wheel tags provide. Individual wheels
```

emmatyping-nv commented on peps/pep-0815.rst:
I think to tie in the last sentence better it'd be good to discuss that variant properties have a priority earlier in the paragraph.

mgorny replied:
https://github.com/wheelnext/peps/pull/11

emmatyping-nv commented on peps/pep-0815.rst:
This has likely been bikeshed already while I was out, so please ignore me if it has. I think "install-time" and "ahead of time" are rather lengthy descriptors, and "install-time" may be a bit of a misnomer. If a user is e.g. resolving a lock file, there may not be any installation.

Perhaps static/dynamic would be clearer?

mgorny replied:
That's a good question. We had static/dynamic for different kinds of plugin API earlier, but that's no longer the case, so it may work.

mgorny replied:
As discussed on Slack, we'll stay with the current terminology for the time being to avoid updating the metadata format now. We may revisit this after the DPO thread.

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
embedded in the wheel; it can be either provided directly by the
wheel maintainer or queried at wheel build time from an AoT plugin. Both kinds
```

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
of plugins are usually implemented as Python packages external
to the resolver which implement the APIs specified in this document.
```

This is really a nit, but "third-party" is unclear as to what the packages are third-party to.

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
variant resolution without explicit user opt-in. Provider packages
```

emmatyping-nv commented on peps/pep-0815.rst:
Should we perhaps mention package signing/attestations here?

mgorny replied:
Not sure how that helps (though I suppose that's related to us inadequately describing the problem). I think the main problem is the potential of arbitrary packages being installed and executed at install time. Attestations don't really change that — packages still specify arbitrary plugins that could compromise your system, whether they are signed or not. On top of that, the security benefit from attestations is disputable at best.

emmatyping-nv replied:
Well, users that want install-time plugins but care about not allowing arbitrary plugins could use attestations to verify that they are only allowing plugins from trusted sources.

mgorny replied:
Ah, so presumably rather than opting in to specific plugin by name, they'd opt in to it by GitHub repository URL, therefore removing the narrow part of attack scope where attacker takes control of the PyPI project but not the GitHub repository.

emmatyping-nv replied:
I think attestations will become much more useful as they expand beyond the current GitHub/GitLab/etc. sources. We've also already seen a number of phishing campaigns against PyPI so I think compromising a PyPI account but not a GitHub account/other attestation source is entirely reasonable.

emmatyping-nv commented on peps/pep-0815.rst:
I think the opening of this paragraph is a bit confusing to me because it is unclear who is should vendor/reimplement/lock plugins. Maintainers? Resolvers?

I presume it is resolvers, but it'd be good to clarify that.

mgorny replied:
The original answer would be "installers", but now you got me thinking whether that's an appropriate term. That said, I don't think "resolvers" is entirely clear either. I suppose we're basically talking about any tools that may need to query the supported variant properties of user's system.

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
Summary of changes
```

Since the last section is titled "Overview" perhaps this section should be a summary of changes?

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
label component. The label must adhere to the following rules:
```

emmatyping-nv commented on peps/pep-0815.rst:
I believe it would be better to use `\w` and `\d` to describe the characters allowed, for internationalization. (Same for the regex below)

mgorny replied:
We deliberately didn't want to open that can of worms. With internationalization, case sensitivity becomes much more a technical issue (think of the Turkish i/I).

emmatyping-nv replied:
I still strongly believe that we should support non-ASCII characters. I suppose case sensitivity is theoretically an issue for wheel project names as well, but practically speaking I don't think case sensitivity has really been an issue. If we care deeply about that, we could allow non-ASCII characters and specify an additional requirement that all labels must not conflict on case-insensitive filesystems.

mgorny replied:
Ah, sorry, you were talking of variant labels and I was thinking of variant properties.

To be honest, I didn't even realize Unicode is permitted in wheel filenames. That said, given that [project names are limited to ASCII](https://packaging.python.org/en/latest/specifications/name-normalization/#name-format) and other filename components are unlikely to contain non-ASCII characters (even if the spec permits that), I'm not convinced there's much of a point in allowing more here.

mgorny replied:
Oh, and just to be clear I'm not opposed to Unicode per se (though e.g. with package names it would be dangerous). My point is rather that I believe the reasons for restrictions we have now (e.g. lowercase) are good, and allowing Unicode would be inconsistent with simultaneously having such a rigorous restrictions.

emmatyping-nv commented on peps/pep-0815.rst:
> same for all wheels using the same label within a single package version

Shouldn't the label be the same globally? It would be very confusing if different packages have different labels for the same sets of variant properties.

mgorny replied:
It's impossible to guarantee globally matching labels.

emmatyping-nv replied:
I agree, but I think it would still be useful to indicate that labels should be globally the same, at least within an ecosystem. Or at least within a project!

I also think this is something indexes could check.

mgorny replied:
Not necessarily. A project that has a single CUDA variant would be clearer with a `-cuda` label, whereas one with variants matching multiple CUDA versions needs e.g. `-cu120`, `-cu125`. By requiring consistent labels, you're basically expecting people to guess up front what labels people may need in the future.

emmatyping-nv commented on peps/pep-0815.rst:
Maybe covered later, but the behavior for what to do when there *isn't* a non-variant wheel should be either specified or explicitly left unspecified.

mgorny replied:
It's not but that's a good idea indeed. In fact, we probably have two cases to cover:

1. There is no non-variant wheel for the newest version, so the installer falls back to an earlier version — we should probably warn about it.
2. There is no non-variant wheel at all — not sure if we need to say this out loud, but the installer should provide a clear error message.

A semi-related question is how the installer should behave if none of the variants are supported.

mgorny replied:
https://github.com/wheelnext/peps/pull/11

emmatyping-nv commented on peps/pep-0815.rst:
Same as above when discussing the label, we should use `\w` and `\d` to describe valid entries. (same for the following two paragraphs)

emmatyping-nv commented on peps/pep-0815.rst:
I noticed that `.` can only be used in the feature value. Is there a reason to exclude it from the names and namespaces? If so it would be good to clarify why above these paragraphs.

mgorny replied:
It's in the rationale section.

https://github.com/wheelnext/peps/pull/6/files#diff-a93e8d47553464b0c59bf04ab34d9c3fad604f10e78674f37048771c2a78e9e4R701-R704

emmatyping-nv commented on peps/pep-0815.rst:
I think multi-value might be something the plugin specifies? But I don't see mention of how a feature gets marked multi-value by the provider.

mgorny replied:
It's in the plugin API.

emmatyping-nv replied:
```suggestion
If a feature is marked as “multi-value” by the provider plugin, a single
variant wheel may define multiple properties sharing the same namespace
and feature name. Otherwise, only a single value can correspond to a
```

Thanks for the  clarification!

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
containing file. Its individual keys are sub-dictionaries that are
```

emmatyping-nv commented on peps/pep-0815.rst:
I think it is better to keep the JSON Schema in the appendix, it would break the flow to have a large JSON document inlined.

mgorny replied:
In the comment, I meant that the `.json` file doesn't get copied to `build/`.

emmatyping-nv replied:
Aha! I would add an RST appendix, link to that at the end, and include the JSON document from the appendix.

mgorny replied:
That still doesn't seem to actually include the raw `.json` file, which we want for people to use in the URL.

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
All the fields are optional, with the following exceptions:
```

emmatyping-nv commented on peps/pep-0815.rst:
```suggestion
:ref:`packaging:wheel-file-name-spec` of the Binary Distribution Format
```

emmatyping-nv commented on peps/pep-0815.rst:
This definitely needs to be clarified. We should specify a `data-dist-info-variants` marker attribute to be added to the `a` tag linking to the variants file for the HTML encoding containing it's hash. We should introduce a new top-level `[dist-info]-variants` key in the JSON simple API project detail dictionary.

emmatyping-nv reviewed (commented):
First off, apologies for so many comments, I can be pedantic at times :)
Second, I've only made it halfway through the specification section, I will likely return to this tomorrow, but wanted to send out some of my comments.

Here's some high level feedback overall:

The PEP does a fantastic job describing the variant properties and is clear when discussing how variants interact with each persona, no small feat!

It'd be good to differentiate installer from resolver and be consistent about that throughout the text.

In the overview of the specification, there is a lot of discussion of mitigations and security measures that should be taken, but no discussion of what the security risks are and what the threat model is. People will want to see an acknowledgement of the risks, probably more than the current Security implications section.

I left a few comments about specifications requiring ASCII. I think we should allow non-ASCII labels and variant properties.

There are a lot of moving parts. It might be good to make and link to a tutorial of how to write a variant package.

The specifications of the formats and resolution algorithms are great!

rgommers commented:
Very nice and detailed review @emmatyping-nv. One remark on it:
 
> It'd be good to differentiate installer from resolver and be consistent about that throughout the text.

This doesn't look like a good idea to me. A resolver isn't a standalone tool/package/thing from a packaging perspective, it's a component inside an installer. The relevant tools are "build backend", "build frontend", "installer". The term "resolver" doesn't seem used at all right now except for in a single section under Rejected Ideas, which seems fine.

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+========
+
+The Python wheel packaging format uses
+:doc:`packaging:specifications/platform-compatibility-tags` to specify
```
The other PEPs don't use the specification prefix

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+- A **variant provider plugin** interface that allows installers to dynamically detect
+  platform attributes and select the most suitable wheel.
+
+The goal is to simplify the user experience to a familiar
```
The abstract is already long, I'd cut the paragraph below

mgorny replied:
As in cut if after the command, or remove the paragraph entirely?

konstin replied:
The whole paragraph, but maybe the "The goal is to simplify the user experience to a familiar `[uv] pip install <package>`" is helpful to keep, given it's the same as in the pytorch quote.

mgorny replied:
Actually, I'll try to rewrite the abstract. I see some unnecessary detail here that should really be moved into other sections, as well as missing bits that we could shortly mention.

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+a significant portion of respondents over the last years have been
+successively using Python for scientific computing purposes, covering
+such areas as Data analysis (steadily over 40% respondents), Machine
+learning (grown to 40% in 2024), Data engineering (around 30%), and
```
nit: Drop the "many more", three examples is good

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+``manylinux_2_34_x86_64`` now implicitly requires ``x86_64-v2`` with no
+support for other x86 version. This lack of support results in
+complexity in managing platform-specific dependencies and compatibility.
+That complexity affects the installation process for users and increases
```
The problem here is really that it's impossible to express any or different x86_64 versions, so you have to pick an oldest version you support and ship only that.

mgorny replied:
I find it hard to understand your point. FWIU the issue there is that packages implicitly require v2 and we can't express that. Being able to ship multiple wheels is tangential to that.

konstin replied:
For most projects, runtime dispatch is out of scope, but they can set `-march=x86-64-v2` (or equivalent). Some projects would like to ship a `-march=x86-64-v4` wheel for the free perf improvement from auto-vectorization, but can't cause they can't tag it. There's not really complexity in managing that as projects currently just can't use that perf feature they'd like to use.

mgorny replied:
Do you think there's a point in mentioning the manylinux issue at all?

konstin replied:
Not really, I don't find it representative of the main problem, especially given that the two reported cases were one user that solved it be updating the VM software and another case was a CPU from 2010 (it looks like the last non-SSE2 CPUs were shipped in 2013).

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+`archspec <https://github.com/archspec/archspec>`__’s authors and
+maintainers, the ability to optimize a package for a specific
+architecture can lead to significant performance improvement (and often
+also reduce power requirements, etc.).
```
```suggestion
As illustrated by
`archspec <https://github.com/archspec/archspec>`__, the ability to
optimize a package for a specific architecture can lead to a
significant performance improvement.
```

konstin commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2282 @@
+          slightly.  The last two bars are labeled "skylake_avx512" and
+          "cascadelake" (both "AVX512") and reach almost 2.0 ns/day.
+
+    GROMACS AVX512 Benchmark
```
What's the GROMACS benchmark, and what do the numbers refer to? The figure description should contain a concise description of what is being measured, just enough for a Python programmer who hasn't heard GROMACS before to understand the impact.

mgorny replied:
@DEKHTIARJonathan

konstin commented on peps/pep-0815.rst:
I'd reduce those sentences (e.g. omit this paragraph), I think the use cases and quotes speak for themselves and make a stronger point.

konstin commented on peps/pep-0815.rst:
We don't have any references for graphics or gaming, so I'd omit that use case

konstin commented on peps/pep-0815.rst:
```suggesion
   (e.g., ``cp313``, ``abi3``, ``none``)
```

Ideally more packages use the stable over the unstable API

konstin commented on peps/pep-0815.rst:
These seem somewhat redundant with the way more concrete examples in the next paragraph, should we merge them?

mgorny replied:
Yeah, let's remove this. No point in listing abstract use cases without examples when we have specific examples.

konstin commented on peps/pep-0815.rst:
Should this be `turn` (without s)?

mgorny replied:
Yes. That said, I'll rewrite this, since it's technically incorrect, given runtime dispatching possibility.

mgorny committed:
Apply suggestions from code review

Co-authored-by: Emma Smith <emsmith@nvidia.com>

konstin commented on peps/pep-0815.rst:
```suggestion
functionality and performance. The current wheel format is not able to
```

konstin commented on peps/pep-0815.rst:
Maybe:

> Wheel variants can capture GPU compatibility in wheel filenames and select the correct wheel at install time.

mgorny replied:
Not really "filenames", but otherwise yes.

konstin commented on peps/pep-0815.rst:
I'd drop these three paragraphs, this information should emerge from the description of the workarounds, and should otherwise already be understood by the PEP reader.

mgorny committed:
Provide examples early in the motivation section

Per https://github.com/wheelnext/peps/pull/6#discussion_r2579567870.

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny commented:
I've already replied to specific comments, but I think these two points are worth discussing top-level:

> It'd be good to differentiate installer from resolver and be consistent about that throughout the text.

I think the PEP as it stands right now roughly splits tools into "installers" and "everything else". The boundary between the two is a bit blurry, and I don't think we're really following it consistently.

The primary idea behind "installers" is that these are packages that need to be able to determine whether variant wheels are supported, and may be executed with elevated privileges. Therefore, from the security PoV there is a risk that variant wheels would require installing and running arbitrary packages with these privileges. The "everything else" is basically tools where we don't worry about that — e.g. tools that only need to process wheels without determining whether they are supported, or build tools that install and run arbitrary packages anyway.

I suppose "installers" is a bit of a misnomer, since there are technically other tools that need to determine whether wheels can be installed. However, I don't think "resolver" is a good replacement. I mean, saying that "resolver should vendor a plugin" is weird — the installer does that, resolver is merely one of its components. Furthermore, the same applies to cases where resolver isn't really used, e.g. when installing a wheel from the local filesystem.

I'm open to improving things, but I don't really have a better term than "installer", and I'm not sure if using a strictly more correct term won't make things harder to comprehend.

> In the overview of the specification, there is a lot of discussion of mitigations and security measures that should be taken, but no discussion of what the security risks are and what the threat model is. People will want to see an acknowledgement of the risks, probably more than the current Security implications section.

Indeed. Probably the "security considerations" section needs to be rewritten to make the problem clearer. I wonder if it would make sense to move it prior to specification; describe the problem there, then the measures taken in the specification part.

emmatyping-nv commented:
> A resolver isn't a standalone tool/package/thing from a packaging perspective, it's a component inside an installer. The relevant tools are "build backend", "build frontend", "installer".

Since resolvers are part of other tools like lockfile generators (e.g. pip-tools), I view resolvers as a component shared between installers and lockfile generators. Resolvers are used elsewhere too (e.g. tools to check licenses of various dependencies, etc.).

> I think the PEP as it stands right now roughly splits tools into "installers" and "everything else". The boundary between the two is a bit blurry, and I don't think we're really following it consistently.

Yeah I think my main point is that the PEP makes specifications about package resolution, but tools other than installers do that.

>  However, I don't think "resolver" is a good replacement. I mean, saying that "resolver should vendor a plugin" is weird — the installer does that, resolver is merely one of its components.

One thing we could do is define "resolvers" as "tools which has a package resolution step". Or even use "resolver tools" (but maybe that is verbose).

> Indeed. Probably the "security considerations" section needs to be rewritten to make the problem clearer. I wonder if it would make sense to move it prior to specification; describe the problem there, then the measures taken in the specification part.

I agree this seems like an improvement! Generally people don't move the sections around, but I think that should be fine since it is important to the flow of the PEP. We could also have a security section in the rationale I suppose.

mgorny committed:
Apply suggestion from @emmatyping-nv

Co-authored-by: Emma Smith <emsmith@nvidia.com>

mgorny committed:
Try to clarify variant ordering in overview

https://github.com/wheelnext/peps/pull/6#discussion_r2586892212

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Suggestions from @rgommers (#8)

* Update the bit about SciPy

Thanks to @rgommers in
https://github.com/wheelnext/peps/pull/6#discussion_r2585536203.

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Sync quotes from https://github.com/wheelnext/wheelnext/pull/110

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Clarify behavior for no non-variant wheels

https://github.com/wheelnext/peps/pull/6#discussion_r2586980105

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny commented:
I'll prepare an update for the security-related stuff tomorrow.

konstin committed:
Improve the FlashAttention section

Don't use tool-specific option, mention the caching problem and better describe why it's no build isolation.

konstin commented on peps/pep-0815.rst:
I would replace this with it being error prone. At least I haven't heard complaints about this being too much effort to document, but i've seen many cases in which those custom workarounds broke, from the pytorch index missing hashes over nvidia packages downloading the wrong artifact on the tooling side to so many complaints about installing GPU packages on the user side.

konstin commented on peps/pep-0815.rst:
We haven't introduces SMs at this point

konstin committed:
Replace non-breaking spaces with regular ones

The regular spaces are ASCII characters, while the non-breaking spaces are special Unicode characters. There's no need in our text to use non-breaking spaces, such as there would be with SI units.

konstin commented on peps/pep-0815.rst:
```suggestion
      // The schema URL will be replaced with the final URL on packaging.python.org
```

mgorny committed:
Apply suggestion from @mgorny

mgorny committed:
Move "Security implications" before the "Specification"

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Expand on the risks and attack scenarios

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Bump pygments requirement

Not sure if we should do that as part of the PEP submission, but we are
using lexers available since 2.19.0, as @konstin noticed.

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Explain install-time/AoT plugins in detail in rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Fix a few conversion errors

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Apply suggestion from @emmatyping-nv

Co-authored-by: Emma Smith <emsmith@nvidia.com>

mgorny committed:
Avoid pip-specific options

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Try to emphasize vendoring plugins more

Signed-off-by: Michał Górny <mgorny@quansight.com>

konstin commented on peps/pep-0815.rst:
We treat the variants json file like an archive file (not like a `.metadata` file), it's a regular top level enty, including having a hash. I'll file a PR.

konstin committed:
Clarify how variants.json is presented on an index

konstin reviewed (commented):
Adding some notes, filed PRs for some others.

konstin committed:
Update peps/pep-0815.rst

Co-authored-by: Emma Smith <emsmith@nvidia.com>

mgorny committed:
Apply suggestion from @emmatyping-nv

Co-authored-by: Emma Smith <emsmith@nvidia.com>

mgorny committed:
Apply suggestion from @konstin

Co-authored-by: konsti <konstin@mailbox.org>

mgorny committed:
Fix trailing whitespace

Signed-off-by: Michał Górny <mgorny@quansight.com>

konstin committed:
Put more focus on torch (#12)

* Put more focus on torch

* Update peps/pep-0815.rst

Co-authored-by: Michał Górny <mgorny@gentoo.org>

* Update peps/pep-0815.rst

Co-authored-by: Michał Górny <mgorny@gentoo.org>

* Update peps/pep-0815.rst

Co-authored-by: Michał Górny <mgorny@gentoo.org>

---------

Co-authored-by: Michał Górny <mgorny@gentoo.org>

mgorny committed:
Drop "and more" from use examples

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove unnecessary paragraph below GROMACS benchmark

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove graphics/gaming case

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Include more wheel tag examples

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove redundant "user stories"

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Correct CPU instruction paragraph

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Improve the GPU paragraph

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove intro to "current workarounds..."

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Replace documentation burden with error-proneness

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Add a note what SMs are

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Replace the schema URL TODO

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Apply suggestions by @konstin

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Add a quote from PyTorch maintainers

Fixes: https://github.com/wheelnext/wheelnext/issues/111
Signed-off-by: Michał Górny <mgorny@quansight.com>

konstin committed:
Line wrapping

mgorny committed:
Expand wheel tag descriptions to match detail from abstract

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Shorten the abstract a bit

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Replace the x86-64 level paragraph

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Update Charlie's e-mail address

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Make the beginning of motivation more illustrative

Attempting to address https://github.com/wheelnext/peps/pull/6#discussion_r2579567870

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Rewrite the introduction to spec

Remove the weird "packaging ecosystem version" and make it clearer
what we're talking about.

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Clarify kinds of tools in the security part of overview

https://github.com/wheelnext/peps/pull/6#discussion_r2586946562

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Copy GROMACS chart desc from the paper

Signed-off-by: Michał Górny <mgorny@quansight.com>

rgommers commented on peps/pep-0815.rst:
I'm not sure who wrote this, but it doesn't look quite right to me. Right now the install widget has 5 flavors, 3 CUDA, 1 ROCm, 1 CPU. No ROCm 7.x, no XPU.

Also, if there were XPU, it should say "three GPU plugins" below, rather than two. 

I think we want to go with the 5 flavors visible in the widget. @atalman do you agree?

mgorny replied:
@atalman did: https://github.com/wheelnext/wheelnext/commit/0a9689b6106f513edc7e2ae0fb2df2fbe10cd3f3#diff-638d2b17e710403ab501eba18ab53c80a281ad637cb7065cd70623354e770f9aL143-R144

dvrogozh replied:
Pytorch builds XPU wheels and publishes them at (https://download.pytorch.org/whl/xpu). However, availability of XPU builds is not yet announced on the pytorch getting started page (https://pytorch.org/get-started/locally/), we are in the process to get XPU there. Links to the XPU downloads are published on the XPU specific pytorch page - https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html.

rgommers commented on peps/pep-0815.rst:
This will need a bit more detail, since the point of this section is to guide the reader to a reference implementation. Maybe link to a gist with the key details (how to install or use the functionality, pointers to the key PRs or parts of the code base)?

konstin replied:
https://github.com/wheelnext/peps/pull/24 - The PR description contains an overview of the current state.

rgommers committed:
Some copy-edits to the text for clarity & project names

konstin committed:
Add a link to the uv reference implementation branch

https://github.com/wheelnext/peps/pull/6#discussion_r2598908698

DEKHTIARJonathan committed:
Acknowledgments Section

DEKHTIARJonathan commented:
I just added a beginning of "acknowledgements" section. I'll keep it updated.
Let me know if I forgot anyone

dvrogozh commented on peps/pep-0815.rst:
> Also, if there were XPU, it should say "three GPU plugins" below, rather than two.

To the point from @rgommers above, I think it will be easier to drop specific number of GPU plugins which needs to be implemented as that's implementation details.

```suggestion
This setup could be improved using GPU plugins that query the
```

rgommers replied:
Thanks @dvrogozh. This is being taken along in gh-25.

DEKHTIARJonathan committed:
Skip build if branch is `norender*`

DEKHTIARJonathan committed:
Linter Fixes

DEKHTIARJonathan committed:
Authors += "Andy R Terrel"

DEKHTIARJonathan committed:
JSON Schema Appendix

mgorny committed:
Updates on full reread (#25)

* Do a grammar run up to "Motivation" section

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Extend "variant properties system" slightly

Add an introduction on what they are, and mention that they are stored
in the metadata. This makes the text easier to process when you are
reading top-to-bottom, given that variant properties haven't been
introduced yet.

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Rewrite the null variant rationale

Rewrite the rationale null variant for better readability and
consistency.  Remove redundant half-paragraph.  Rewrite the sentences
that sound like specification (specification already covers these bits).

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Clarify the precedence between null variant and non-variant wheels

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Updates up to "security implications"

Signed-off-by: Michał Górny <mgorny@quansight.com>

* First part of specification update

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Final touches to the specification

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Final sections, finally!

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Apply suggestions from code review

Co-authored-by: konsti <konstin@mailbox.org>

* Apply more suggestions from Konsti

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Restore "a choice"

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Apply suggestion from @rgommers

Co-authored-by: Ralf Gommers <ralf.gommers@gmail.com>

* "could leads" → "leads"

Signed-off-by: Michał Górny <mgorny@quansight.com>

---------

Signed-off-by: Michał Górny <mgorny@quansight.com>
Co-authored-by: konsti <konstin@mailbox.org>
Co-authored-by: Ralf Gommers <ralf.gommers@gmail.com>

mgorny committed:
Try a more smooth title (#27)

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Add more links and cross-references (#28)

* Add more links and cross-references

Add links to projects we're talking about, when we're referencing them
for the first time or haven't mentioned them in a while.  Add more
cross-references to make finding relevant bits easier, especially if
we're talking about something that's specified later on.  Cross-link to
specifications more.

Signed-off-by: Michał Górny <mgorny@quansight.com>

* Undo OpenTeams link

Signed-off-by: Michał Górny <mgorny@quansight.com>

---------

Signed-off-by: Michał Górny <mgorny@quansight.com>

DEKHTIARJonathan commented:
I can't approve my own PR - But thumbs up for me !

warsawnv commented on peps/pep-0815.rst:
```diff
@@ -0,0 +1,2494 @@
+PEP: 815
```
```suggestion
PEP: 817
```

DEKHTIARJonathan replied:
I'll merge this once the PR is merged I need to rename a few files ;)

atalman reviewed (approved):
lgtm

warsawnv reviewed (approved):
This should be PEP 817 (the next available number as of this writing).

Also, you should add at least one (if not all) core devs in the authors list to `.github/CODEOWNERS`.

DEKHTIARJonathan commented:
> Also, you should add at least one (if not all) core devs in the authors list to .github/CODEOWNERS.
Ah good I didn't know, will do

seemethere reviewed (approved):
LGTM!

rgommers reviewed (approved):
LGTM too!

DEKHTIARJonathan commented:
All authors have approved - I'll merge !

Congrats everybody !
